README
tibae module for Qore
version 0.1 by David Nichols

The "tibae" module provides TIBCO Adapter support; it allows qore programs to send and receive messages to and from TIBCO Adapters (or implement a TIBCO Adapter in qore)  based on an external repository.

This module provides the TibcoAdapter class and associated constants and helper functions.

Note that early versions of these release notes stated that it was not possible to make an adapter that sent synchronous replies - this is incorrect, just set the subject to the replySubject key of the incoming message hash which should be an inbox (point-to-point) subject name.
* Linux: use the compiler required by the SDK version
* Solaris: qore must be compiled with the SunPro CC compiler in order to be able to use the "tibae" module.
* OS X: There is finally Rendezvous support for OS X, so hopefully they will support the Adapter SDK sometime soon

TIBCO ActiveEnterprise(TM) is a trademark of TIBCO, Inc.

Documentation in docs/tibco-module-doc.html
(built from docbook sources in docs/docbook/*)

test/example scripts: test/tibae-test.q test/tibae-listener.q


BUILD AND INSTALL
-----------------

Note that this module must be built with the compiler required by the TIBCO Adapter SDK in order for the "tibae" module to be built.

You must have both TIBCO Rendezvous and the AE SDK installed (along with the supported C++ compiler) before you can build this module.

Make sure that the:
*) RV_ROOT
*) SDK_ROOT
*) TPCL_ROOT and
*) JMS_ROOT
environment variables are pointing to your Rendezvous, SDK, TPCL, and EMS directories respectively before calling configure (JMS_ROOT only required for SDK 5.5 and above).

Otherwise you can use the --with-tibrv, --with-tibae, with-tibae-tpcl, and --with-tibco-ems configure options.  The "tibae" module will compile with SDK 4.* versions, but there are so many bugs in this version of the SDK (including some horrible dynamic memory leaks) that it doesn't make sense to use anything before 5.2.1...
Note that newer SDKs may work with HP-UX PA-RISC 11.*, so the restriction on building this module on HP-UX PA-RISC has been removed from the configure script.

configure
make
sudo make install

The configure script will find out where your qore module directory is found and set this for the install directory.

BUILD FROM SVN
--------------

reconf.sh
configure
make
sudo make install


ChangeLog
2010-06-21 09:12  david_nichols

	* qore-tibae-module.spec: updated spec files to generically
	  determine *suse versions for distribution tagging in rpm file
	  names

2010-06-20 06:03  david_nichols

	* ChangeLog: fixed ChangeLogs with non-broken svn2cl

2010-06-14 14:05  david_nichols

	* docs/tibae-module-doc.html: updated docs for new release

2010-06-14 14:04  david_nichols

	* docs/docbook/get-docbook-root.sh: added missing file

2010-06-10 14:22  david_nichols

	* docs/docbook/Makefile, docs/docbook/tibae-module.docbook,
	  docs/qore-style.css, docs/tibae-module-doc.html: updated docs for
	  new release

2010-05-12 11:49  david_nichols

	* configure.ac: solaris and sun studio compile and warning
	  fixes/enhancements

2010-05-11 14:44  david_nichols

	* src/Makefile.am: compile fixes for solaris

2010-05-08 17:26  david_nichols

	* RELEASE-NOTES, src/tibae.cc: updated conversions to use more
	  efficient APIs; updated release notes

2010-05-08 16:57  david_nichols

	* RELEASE-NOTES, configure.ac, qore-tibae-module.spec,
	  src/QoreApp.cc, src/tibae-module.cc, src/tibae.cc: updated to new
	  date/time APIs in qore 0.8.0; updated release to 1.0.4

2009-12-09 12:58  david_nichols

	* configure.ac: fixed all modules to use decimal comparisons with
	  the qore version string instead of octal so that they will still
	  compile with qore v0.8.0 and up...

2009-11-22 06:03  david_nichols

	* configure.ac: updated configure to assume 64-bit build if
	  host_cpu = "amd64"

2009-10-30 08:22  david_nichols

	* configure.ac: updated all module configure scripts to use gnutar
	  on darwin to avoid picking up OSX resource files in source
	  releases

2009-10-05 08:53  david_nichols

	* ChangeLog: udpated changelog

2009-09-21 11:07  david_nichols

	* configure.ac: updated module configure scripts to select 64-bit
	  target mode by default on snow leopard i386

2009-09-10 06:49  david_nichols

	* configure.ac: updated configure.ac to properly compile in 32-bit
	  mode on OSX snow leopard (10.6)

2009-08-27 07:30  david_nichols

	* reconf.sh: fixed reconf to work on platforms with broken which
	  (like solaris)

2009-08-26 11:53  david_nichols

	* RELEASE-NOTES, configure.ac, qore-tibae-module.spec,
	  src/QoreApp.cc: fixed memory leaks in
	  TibcoAdapter::sendSubjectWithSyncReply() when an error
	  occurs creating the MPublisher
	  
	  fixed other memory leaks in exception handling in TIBCO message
	  processing by
	  using std::auto_ptr to handle memory deallocations automatically

2009-07-20 13:16  david_nichols

	* configure.ac, qore-tibae-module.spec: added opensuse 11.2
	  detection to spec file, fixed rhel detection in spec files,
	  updated configure.ac to work with older versions of autotools

2009-06-08 08:38  david_nichols

	* qore-tibae-module.spec: fixed qore module dir detection

2009-06-02 05:24  david_nichols

	* qore-tibae-module.spec: fixed typo in sles version detection

2009-01-08 18:43  david_nichols

	* qore-tibae-module.spec: simplified spec files a little

2009-01-08 15:13  david_nichols

	* qore-tibae-module.spec: commented-out additional "Source0" line
	  so that packages will build on Fedora 10 (OBS)

2009-01-08 14:44  david_nichols

	* reconf.sh: updated reconf.sh to look for glibtoolize as well as
	  libtoolize

2009-01-07 13:34  david_nichols

	* RELEASE-NOTES: updated release notes

2009-01-07 13:32  david_nichols

	* Makefile.am, configure.ac, ltmain.sh, qore-tibae-module.spec,
	  reconf.sh: upasdted to version 1.0.2; updated reconf.sh to use
	  libtoolize to generate ltmain.sh

2008-11-03 15:17  david_nichols

	* qore-tibae-module.spec: fixed version # in spec file

2008-11-03 15:17  david_nichols

	* ChangeLog, Makefile.am, RELEASE-NOTES, configure.ac,
	  qore-tibae-module.spec, src/Makefile.am, src/QC_TibcoAdapter.cc,
	  src/QoreApp.cc, src/TibCommandLine.cc, src/tibae-module.cc,
	  src/tibae-module.h, src/tibae.cc: updated to new module handling;
	  updated version to 1.0.1

2008-10-30 17:20  david_nichols

	* qore-tibae-module.spec: various fixes

2008-10-23 06:30  david_nichols

	* configure.ac: updated to set 64-bit compile mode by default on
	  itanium with g++ and aCC

2008-10-08 20:16  david_nichols

	* qore-tibae-module.spec, src/Makefile.am: spec file and module
	  build fixes using prefix=/usr and make install
	  DESTDIR=$RPM_BUILD_ROOT

2008-10-07 18:12  david_nichols

	* ChangeLog: updated changelog

2008-10-07 18:12  david_nichols

	* README, configure.ac, test/tibae-listener.q, test/tibae-test.q:
	  fixed configure, fixed test scripts to look for adapter file in
	  the same directory as the script

2008-10-07 06:20  david_nichols

	* ChangeLog: updated changelog

2008-10-01 20:02  david_nichols

	* src/Makefile.am: fixed makefile for qore cppflags

2008-09-16 20:54  david_nichols

	* configure.ac: build and configure fixes

2008-09-14 07:21  david_nichols

	* Makefile.am, qore-tibae-module.spec: fixed tibae packaging to
	  include all test scripts and test adapter

2008-09-14 07:16  david_nichols

	* test/new.dat: moved test adapter to tibae project

2008-09-09 07:20  david_nichols

	* ChangeLog, qore-tibae-module.spec: updated license in spec file
	  to LGPL, updated ChangeLog

2008-09-08 17:45  david_nichols

	* Makefile.am, qore-tibae-module.spec: fixed rpm packaging for docs

2008-09-06 07:08  david_nichols

	* Makefile.am, qore-tibae-module.spec: added documentation

2008-09-05 13:47  david_nichols

	* README, configure.ac, docs, docs/docbook, docs/docbook/Makefile,
	  docs/docbook/bookinfo.docbook, docs/docbook/driver.xsl,
	  docs/docbook/module.docbook, docs/docbook/tibae-module.docbook,
	  docs/img, docs/img/acrobat.gif, docs/img/acrobat_a.gif,
	  docs/img/disk.gif, docs/img/qbullet-note.gif,
	  docs/img/qbullet-note_a.gif, docs/img/remote.gif,
	  docs/img/remote_a.gif, docs/img/scrollup.gif,
	  docs/img/scrollup_a.gif, docs/img/word.gif, docs/qore-style.css,
	  docs/tibae-module-doc.html, qore-tibae-module.spec,

src/tibae.cc
/*
  TIBCO/tibae.cc

  TIBCO Active Enterprise integration to QORE

  Copyright 2003 - 2008 David Nichols

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "tibae-module.h"

#include "tibae.h"
#include "TibCommandLine.h"
#include "QoreApp.h"

#include <memory>

static inline class AbstractQoreNode *map_minstance_to_node(const MInstance *min, ExceptionSink *xsink)
{
   std::auto_ptr<MEnumerator<MString, MData *> >me(min->newEnumerator());   
   MString name;
   MData *val;

   ReferenceHolder<QoreHashNode> h(new QoreHashNode(), xsink);
   while (me->next(name, val)) {
      h->setKeyValue((char *)name.c_str(), map_mdata_to_node(val, xsink), xsink);
      if (*xsink)
	 return 0;
   }

   return h.release();
}

// maps a TIBCO sequence to a QORE list
static inline class AbstractQoreNode *map_msequence_to_node(const MSequence *ms, ExceptionSink *xsink)
{
   ReferenceHolder<QoreListNode> rv(new QoreListNode(), xsink);

   for (unsigned i = 0; i < ms->size() && !xsink->isEvent(); i++)
      rv->push(map_mdata_to_node((MData *)(*ms)[i], xsink));

   if (xsink->isEvent())
      return 0;

   return rv.release();
}

typedef MEnumerator<MString, MData *> ma_enumerator_t;

// maps a TIBCO associative list to a QORE hash
static inline class AbstractQoreNode *map_massoclist_to_node(const MAssocList *mal, ExceptionSink *xsink)
{
   ReferenceHolder<QoreHashNode> h(new QoreHashNode(), xsink);

   std::auto_ptr<ma_enumerator_t> me(mal->newEnumerator());
   MString name;
   MData *val;

   while (me->next(name, val)) {
      h->setKeyValue((char *)name.c_str(), map_mdata_to_node(val, xsink), xsink);
      if (*xsink)
	 return 0;
   }

   return h.release();
}

// maps a TIBCO union to a QORE hash
static inline class AbstractQoreNode *map_munion_to_node(const MUnion *mu, ExceptionSink *xsink)
{
   ReferenceHolder<QoreHashNode> h(new QoreHashNode(), xsink);

   MString MSkey = mu->getMemberName();
   h->setKeyValue((char *)MSkey.c_str(), map_mdata_to_node((MData *)mu->get(MSkey), xsink), xsink);
   
   if (*xsink)
      return 0;

   return h.release();
}

// maps a TIBCO MTree to a QORE node
class AbstractQoreNode *map_mdata_to_node(MData *md, ExceptionSink *xsink)
{
   const MInstance *min;

   // is it an MInstance?
   if ((min = MInstance::downCast(md)))
      return map_minstance_to_node(min, xsink);

   const MAssocList *mal;
   // is it an associative list?
   if ((mal = MAssocList::downCast(md)))
      return map_massoclist_to_node(mal, xsink);

   const MSequence *ms;
   // or is it a sequence?
   if ((ms = MSequence::downCast(md)))
      return map_msequence_to_node(ms, xsink);

   const MUnion *mu;
   // or is it a union?
   if ((mu = MUnion::downCast(md)))
      return map_munion_to_node(mu, xsink);
   
   const MInteger *mi;
   if ((mi = MInteger::downCast(md)))
      return new QoreBigIntNode(mi->getMi8());
   
   const MStringData *msd;
   if ((msd = MStringData::downCast(md)))
   {
      MString ms = msd->getAsString();
      return new QoreStringNode((char *)ms.c_str());
   }
   
   const MReal *mr;
   if ((mr = MReal::downCast(md)))
      return new QoreFloatNode(mr->getAsDouble());
   
   const MDateTime *mdt;
   if ((mdt = MDateTime::downCast(md)))
#ifdef _QORE_HAS_TIME_ZONES
      return DateTimeNode::makeAbsolute(currentTZ(), mdt->getTime(), mdt->getMicroSeconds());
#else
      return new DateTimeNode(mdt->getYear(), mdt->getMonth(), mdt->getDay(), mdt->getHour(), mdt->getMinute(), mdt->getSecond(), mdt->getMicroSeconds() / 1000);
#endif
   
   const MDate *mdate;
   if ((mdate = MDate::downCast(md)))
      return new DateTimeNode(mdate->getYear(), mdate->getMonth(), mdate->getDay());
   
   const MBool *mb;
   if ((mb = MBool::downCast(md)))
      return get_bool_node((bool)mb->getAsBoolean());

   const MBinary *mbin;
   if ((mbin = MBinary::downCast(md)))
   {
      BinaryNode *b = new BinaryNode;
      b->append(mbin->getData(), mbin->size());
      return b;
   }

   const MInterval *mint;
   if ((mint = MInterval::downCast(md)))
      return new DateTimeNode(0, 0, 0, 0, 0, mint->getSeconds(), mint->getMicroSeconds() / 1000, true);

   const MTime *mtime;
   if ((mtime = MTime::downCast(md)))
#ifdef _QORE_HAS_TIME_ZONES
      return DateTimeNode::makeAbsoluteLocal(currentTZ(), mtime->getHour() * 3600 + mtime->getMinute() * 60 + mtime->getSecond(), mtime->getMicroSeconds());      
#else
      return new DateTimeNode(1970, 1, 1, mtime->getHour(), mtime->getMinute(), mtime->getSecond(), mtime->getMicroSeconds() / 1000);
#endif

   xsink->raiseException("MAP-ERROR", "can't map MData element of class '%s' to QORE type", md->getClassName());
   return NULL;
}

void set_properties(MAppProperties *appProperties, const QoreHashNode *h, TibCommandLine &tcl, ExceptionSink *xsink)
{
   QORE_TRACE("set_properties()");

   // variable hash for overridding global variables
   const QoreHashNode *vh = NULL;

   ConstHashIterator hi(h);
   while (hi.next())
   {
      const char *key = hi.getKey();
      if (!hi.getValue())
      {
	 xsink->raiseException("TIBCO-INVALID-PROPERTIES-HASH", "properties hash key '%s' has value = NOTHING",	key);
	 return;
      }

      if (!strcmp(key, "Vars"))
      {
	 vh = dynamic_cast<const QoreHashNode *>(hi.getValue());
	 if (vh)
	    continue;
      }

      const AbstractQoreNode *v = hi.getValue();
      const QoreStringNode *str = dynamic_cast<const QoreStringNode *>(v);
      if (!str)
      {
	 xsink->raiseException("TIBCO-INVALID-PROPERTIES-HASH",
			       "properties hash has invalid type '%s' for key '%s' (must be string)",
			       v ? v->getTypeName() : "NOTHING", key);
	 return;
      }
      const char *val = str->getBuffer();

      if (!strcmp(key, "AppVersion"))
	 appProperties->setAppVersion(val);
      else if (!strcmp(key, "AppInfo"))
	 appProperties->setAppInfo(val);
      else if (!strcmp(key, "AppName"))
	 appProperties->setAppName(val);
      else if (!strcmp(key, "RepoURL"))
	 appProperties->setRepoURL(val);
      else if (!strcmp(key, "ConfigURL"))
	 appProperties->setConfigURL(val);
#ifdef DEBUG
      else printe("ignoring properties member %s=%s\n", key, val);
#endif
   }

   if (vh)
   {
      ConstHashIterator vhi(vh);
      while (vhi.next())
      {
	 const char *key = vhi.getKey();
	 if (!key || !key[0])
	    continue;

	 const AbstractQoreNode *n = vhi.getValue();
	 const QoreStringNode *str = dynamic_cast<const QoreStringNode *>(n);
	 if (!str || !str->strlen())
	    continue;
	 //printd(5, "setting override for global variable %s=%s\n", key, str->getBuffer());
	 tcl.add(key, str->getBuffer());
      }
   }

   if (tcl.argc)
      appProperties->setCommandLine(tcl.argc, tcl.argv);

   appProperties->setMultiThreaded(); 
   //appProperties->setDefaultStringEncoding(MEncoding::M_ASCII);
   //appProperties->setDefaultStringEncoding(MEncoding::M_LATIN_1);


}


src/tibae-module.h
/*
  modules/TIBCO/tibco-module.h

  TIBCO integration to QORE

  Copyright 2003 - 2008 David Nichols

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef _QORE_TIBCO_MODULE_H

#define _QORE_TIBCO_MODULE_H

#include "../config.h"

#include <qore/Qore.h>

QoreStringNode *tibae_module_init();
void tibae_module_ns_init(QoreNamespace *rns, QoreNamespace *qns);
void tibae_module_delete();

#endif


src/TibCommandLine.cc
/*
  TibCommandLine.cc

  TIBCO Active Enterprise integration to QORE

  Copyright 2003 - 2008 David Nichols

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "tibae-module.h"

#include "TibCommandLine.h"

#include <stdlib.h>

void TibCommandLine::add_intern(char *str)
{
   // resize array
   if (argc == alloc) {
      alloc += 10;
      argv = (char **)realloc(argv, sizeof(char *) * alloc);
   }
   argv[argc] = str;
   //printd(5, "add_intern %d='%s'\n", argc, str);
   argc++;
}

TibCommandLine::~TibCommandLine()
{
   if (argv)
   {
      // delete every other string starting with index 2
      for (int i = 2; i < argc; i += 2)
	 free(argv[i]);
      free(argv);
   }
}

void TibCommandLine::add(const char *key, const char *val)
{	 
   if (!argc)
      // add a dummy entry for the program name
      add_intern((char *)"dummy");
   QoreString str;
   add_intern((char *)"-system:clientVar");
   str.sprintf("%s=%s", key, val);
   add_intern(str.giveBuffer());
}


src/QC_TibcoAdapter.cc
/*
  QC_TibcoAdapter.cc

  TIBCO Active Enterprise integration to QORE

  Copyright 2003 - 2008 David Nichols

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "tibae-module.h"

#include "QC_TibcoAdapter.h"
#include "TibCommandLine.h"
#include "tibae.h"
#include "QoreApp.h"

#include <memory>

qore_classid_t CID_TIBAE;

// usage: new TibcoAdapter(session-name, properties, classlist, [, service, network, daemon])
void TIBAE_constructor(QoreObject *self, const QoreListNode *params, ExceptionSink *xsink) {
   QORE_TRACE("TIBAE_constructor");

   const char *session_name, *service = NULL, *network = NULL, *daemon = NULL;
   const QoreStringNode *p0;
   const QoreHashNode *p1, *p2;
   const AbstractQoreNode *p;

   if (!(p0 = test_string_param(params, 0)) || !(p1 = test_hash_param(params, 1))) {
      xsink->raiseException("TIBCO-PARAMETER-ERROR", "invalid parameters passed to TibcoAdapter::constructor(), expecting session name (string), properties (object), [classlist (object), service (string), network (string), daemon (string)]");

      return;
   }

   session_name = p0->getBuffer();
   QoreHashNodeHolder classlist(xsink);
   if ((p2 = test_hash_param(params, 2))) {
      // FIXME: check that classlist hash has only String values!
      classlist = p2->QoreHashNode::copy();
   }

   QoreString tmp;
   if ((p = get_param(params, 3))) {
      const QoreStringNode *str = dynamic_cast<const QoreStringNode *>(p);
      if (str)
	 service = str->getBuffer();
      else if (p->getType() == NT_INT) {
	 tmp.sprintf("%lld", reinterpret_cast<const QoreBigIntNode *>(p)->val);
	 service = tmp.getBuffer();
      }
   }

   const QoreStringNode *pstr;
   if ((pstr = test_string_param(params, 3)))
      service = pstr->getBuffer();

   if ((pstr = test_string_param(params, 4)))
      network = pstr->getBuffer();

   if ((pstr = test_string_param(params, 5)))
      daemon = pstr->getBuffer();

   // create adapter instance
   printd(1, "TIBAE_constructor() session=%s service=%s network=%s daemon=%s\n",
	  session_name, 
	  service ? service : "(null)", 
	  network ? network : "(null)", 
	  daemon  ? daemon  : "(null)");
   QoreApp *myQoreApp = NULL;

   MAppProperties *appProps = new MAppProperties();

   TibCommandLine tcl;
   set_properties(appProps, p1, tcl, xsink); 

   if (*xsink)
      return;

   try {
      //printd(5, "before QoreApp constructor\n");
      myQoreApp =
         new QoreApp(appProps, session_name, classlist.release(), service, network, daemon);

      //printd(5, "after QoreApp constructor (%08p)\n", myQoreApp);

      //printd(5, "before start()\n");
      myQoreApp->start(Mfalse);
      //printd(5, "after start()\n");
   }
   catch (MException &te) {
      xsink->raiseException("TIBCO-EXCEPTION", "Exception thrown in Tibco() constructor %s: %s",
			 te.getType().c_str(), te.getDescription().c_str());
      if (myQoreApp)
	 myQoreApp->deref(xsink);

      return;
   }
   self->setPrivate(CID_TIBAE, myQoreApp);
   printd(5, "TIBAE_constructor() this=%08p myQoreApp=%08p\n", self, myQoreApp);

}

void TIBAE_copy(QoreObject *self, QoreObject *old, QoreApp *myQoreApp, class ExceptionSink *xsink) {
   xsink->raiseException("TIBCO-ADAPTER-COPY-ERROR", "copying TibcoAdapter objects is curently not supported");
}

// usage: TIBAE_sendSubject(subject, function_name, message)
static AbstractQoreNode *TIBAE_sendSubject(QoreObject *self, QoreApp *myQoreApp, const QoreListNode *params, class ExceptionSink *xsink) {
   const QoreStringNode *p0, *p1;
   const AbstractQoreNode *p2;

   // check input parameters
   if (!(p0 = test_string_param(params, 0)) ||
       !(p1 = test_string_param(params, 1)) ||
       is_nothing((p2 = get_param(params, 2))))
   {
      xsink->raiseException("TIBCO-SENDSUBJECT-PARAMETER-ERROR", "invalid parameters passed to TibcoAdapter::sendSubject(), expecting subject (string), function name (string), message (hash)");
      return NULL;
   }

   const char *subject = p0->getBuffer();
   const char *fname = p1->getBuffer();

   // try to send message
   try {
      myQoreApp->send(subject, fname, p2, xsink);
   }
   catch(MException &te) {
      xsink->raiseException("TIBCO-EXCEPTION", 
			    "QoreException caught while sending \"%s\" with subject \"%s\": %s: %s", 
			    fname, subject, te.getType().c_str(), te.getDescription().c_str());      
   }
   return 0;
}

// usage: Tibco::sendSubjectWithSyncReply(subject, function_name, message[, timeout])
static AbstractQoreNode *TIBAE_sendSubjectWithSyncReply(QoreObject *self, QoreApp *myQoreApp, const QoreListNode *params, class ExceptionSink *xsink) {
   const QoreStringNode *p0, *p1;
   const AbstractQoreNode *p2, *p3;

   // check input parameters
   if (!(p0 = test_string_param(params, 0)) ||
       !(p1 = test_string_param(params, 1)) ||
       is_nothing((p2 = get_param(params, 2)))) {
      xsink->raiseException("TIBCO-SEND-WITH-SYNC-REPLY-PARAMETER-ERROR", "invalid parameters passed to tibco_send_with_sync_reply(), expecting subject (string), function name (string), message data, [timeout (date/time or integer milliseconds)]");
      return NULL;
   }

   const char *fname = p1->getBuffer();

   // set timeout parameter if present
   int timeout = 0;
   if ((p3 = get_param(params, 3)))
      timeout = getMsZeroInt(p3);

   // try to send message
   try {
      printd(2, "TIBAE_sendSubjectWithSyncReply() calling sendWithSyncReply()\n");
      return myQoreApp->sendWithSyncReply(p0->getBuffer(), fname, p2, timeout, xsink);
   }
   catch(MException &te) {
      xsink->raiseException("TIBCO-EXCEPTION", 
			    "QoreException caught while sending \"%s\" with subject \"%s\": %s: %s", 
			    fname, p0->getBuffer(), te.getType().c_str(), te.getDescription().c_str());      
   }

   return 0;
}

// Tibco::receive(subject, [timeout])
static AbstractQoreNode *TIBAE_receive(QoreObject *self, QoreApp *myQoreApp, const QoreListNode *params, class ExceptionSink *xsink) {
   const QoreStringNode *p0;

   if (!(p0 = test_string_param(params, 0))) {
      xsink->raiseException("TIBCO-RECEIVE-PARAMETER-ERROR", "invalid parameters passed to tibco_receive(), expecting subject (string), [timeout (date/time or integer milliseconds)]");
      return NULL;
   }

   const char *subject = p0->getBuffer();
   unsigned long timeout = 0;

   const AbstractQoreNode *p1; 
   if ((p1 = get_param(params, 1)))
      timeout = getMsZeroInt(p1);
   
   return myQoreApp->receive(subject, timeout, xsink);
}

//------------------------------------------------------------------------------
// Parameters:
// * class name (string), in Tibco docs called "name of the class repository object in which this operation is defined"
//   Example: "greetings".
// * method name (string), in Tibco docs "name of the operation repository that defined this operation"
//   Example: "setGreetings"
// * data (hash), keys are string parameter names + appropriate type (according to repository) of values
// * optional: timeout (date/time or integer, millseconds), default is 60 seconds, 0 means infinite
// * optional: client name (string), in Tibco docs "this name must refer to a client repository object defined 
//   in endpoint.clients directory
//   Default value is ""
//
// Returns hash with data send as a reply to this call.
//
static AbstractQoreNode* TIBAE_operationsCallWithSyncResult(QoreObject* self, QoreApp* myQoreApp, const QoreListNode *params, ExceptionSink *xsink) {
   const char* err = "Invalid parameters. Expected: class name (string), method name (string), data (hash), "
      "[timeout (integer in millis or date/time), ] [client name (string)]";
   const char* func = "TIBCO-OPERATIONS-CALL-WITH-SYNC_RESULT";
   const QoreStringNode* class_name = test_string_param(params, 0);
   if (!class_name) {
      return xsink->raiseException(func, err);
   }
   const char* class_name_extracted = class_name->getBuffer(); 
   const QoreStringNode* method_name = test_string_param(params, 1);
   if (!method_name) {
      return xsink->raiseException(func, err);
   } 
   const char* method_name_extracted = method_name->getBuffer();
   const QoreHashNode *data = test_hash_param(params, 2);
   if (!data) {
      return xsink->raiseException(func, err);
   }
   unsigned timeout = 60 * 1000;
   const char* client_name = "";
  
   int next_item = 3;
   const AbstractQoreNode* n = get_param(params, 3);
   qore_type_t ntype = n ? n->getType() : 0;
   if (ntype == NT_INT) {
      timeout = (unsigned)(reinterpret_cast<const QoreBigIntNode *>(n)->val);
      ++next_item;
   } else if (ntype == NT_DATE) {
      const DateTimeNode *date = reinterpret_cast<const DateTimeNode *>(n);
      timeout = (unsigned)date->getRelativeMilliseconds();
      ++next_item;
   }
   const QoreStringNode *nstr = test_string_param(params, next_item);
   if (nstr) {
      client_name = nstr->getBuffer();
   }

   return myQoreApp->operationsCallWithSyncResult(class_name_extracted, method_name_extracted, data, timeout, client_name, xsink);
}

//------------------------------------------------------------------------------
// The same parameters as for TIBAE_operationsCallWithSyncResult, 
// except for the timeout (not needed), always returns 0.
//
static AbstractQoreNode* TIBAE_operationsOneWayCall(QoreObject* self, QoreApp* myQoreApp, const QoreListNode *params, ExceptionSink *xsink) {
   const char* err = "Invalid parameters. Expected: class name (string), method name (string), data (hash), "
      "[client name (string)]";
   const char* func = "TIBCO-OPERATIONS-ONE-WAY-CALL";
   const QoreStringNode* class_name = test_string_param(params, 0);
   if (!class_name) {
      return xsink->raiseException(func, err);
   }
   const char* class_name_extracted = class_name->getBuffer();
   const QoreStringNode* method_name = test_string_param(params, 1);
   if (!method_name) {
      return xsink->raiseException(func, err);
   }
   const char* method_name_extracted = method_name->getBuffer();
   const QoreHashNode *data = test_hash_param(params, 2);
   if (!data) {
      return xsink->raiseException(func, err);
   }
   const char* client_name = "";

   const QoreStringNode* n = test_string_param(params, 3);
   if (n) {
      client_name = n->getBuffer();
   }

   myQoreApp->operationsOneWayCall(class_name_extracted, method_name_extracted, data, client_name, xsink);
   return 0;
}

// The same parameters as for TIBAE_operationsCallWithSyncResult (including timeout).
// Always return 0. To get the reply use combination of class name + method name passed to this call.
static AbstractQoreNode* TIBAE_operationsAsyncCall(QoreObject* self, QoreApp* myQoreApp, const QoreListNode *params, ExceptionSink *xsink) {
   const char* err = "Invalid parameters. Expected: class name (string), method name (string), data (hash), "
      "[timeout (integer in millis or date/time), ] [client name (string)]";
   const char* func = "TIBCO-OPERATIONS-ASYNC-CALL";
   const QoreStringNode* class_name = test_string_param(params, 0);
   if (!class_name) {
      return xsink->raiseException(func, err);
   }
   const char* class_name_extracted = class_name->getBuffer();
   const QoreStringNode* method_name = test_string_param(params, 1);
   if (!method_name) {
      return xsink->raiseException(func, err);
   }
   const char* method_name_extracted = method_name->getBuffer();
   const QoreHashNode *data = test_hash_param(params, 2);
   if (!data) {
      return xsink->raiseException(func, err);
   }
   unsigned timeout = 60 * 1000;
   const char* client_name = "";

   int next_item = 3;
   const AbstractQoreNode *n = get_param(params, 3);
   qore_type_t ntype = n ? n->getType() : 0;
   if (ntype == NT_INT) {
      timeout = (unsigned)(reinterpret_cast<const QoreBigIntNode *>(n)->val);
      ++next_item;
   } else if (ntype == NT_DATE) {
      const DateTimeNode *date = reinterpret_cast<const DateTimeNode *>(n);
      timeout = (unsigned)date->getRelativeMilliseconds();
      ++next_item;
   }
   const QoreStringNode *nstr = test_string_param(params, next_item);
   if (nstr) {
      client_name = nstr->getBuffer();
   }

   myQoreApp->operationsAsyncCall(class_name_extracted, method_name_extracted, data, timeout, client_name, xsink);
   return 0;
}

//------------------------------------------------------------------------------
// Parameters:
// * class name (string)
// * method name (string)
// The class/method need to be invoked prior with operationsAsyncCall().
//
// Returns hash with retrieved values.
//
static AbstractQoreNode* TIBAE_operationsGetAsyncCallResult(QoreObject* self, QoreApp* myQoreApp, const QoreListNode *params, ExceptionSink *xsink) {
   const char* err = "Invalid parameters. Expected: class name (string), method name (string)";
   const char* func = "TIBCO-OPERATIONS-GET-ASYNC-RESULT";
   const QoreStringNode* class_name = test_string_param(params, 0);
   if (!class_name) {
      return xsink->raiseException(func, err);
   }
   const char* class_name_extracted = class_name->getBuffer();
   const QoreStringNode* method_name = test_string_param(params, 1);
   if (!method_name) {
      return xsink->raiseException(func, err);
   }
   const char* method_name_extracted = method_name->getBuffer();

   return myQoreApp->operationsGetAsyncCallResult(class_name_extracted, method_name_extracted, xsink);
}

QoreClass *initTibcoAdapterClass() {
   QORE_TRACE("initTibcoAdapterClass()");

   QoreClass *QC_TIBAE = new QoreClass("TibcoAdapter", QDOM_NETWORK);
   CID_TIBAE = QC_TIBAE->getID();
   //printd(5, "initTibcoAdapterClass() CID_TIBAE=%d\n", CID_TIBAE);
   QC_TIBAE->setConstructor(TIBAE_constructor);
   QC_TIBAE->setCopy((q_copy_t)TIBAE_copy);
   QC_TIBAE->addMethod("receive",                  (q_method_t)TIBAE_receive);
   QC_TIBAE->addMethod("sendSubject",              (q_method_t)TIBAE_sendSubject);
   QC_TIBAE->addMethod("sendSubjectWithSyncReply", (q_method_t)TIBAE_sendSubjectWithSyncReply);

   // operations
   QC_TIBAE->addMethod("callOperationWithSyncReply",  (q_method_t)TIBAE_operationsCallWithSyncResult);
   QC_TIBAE->addMethod("callOperationOneWay",         (q_method_t)TIBAE_operationsOneWayCall);

   /*
   // commented out for now as the retrieving async call method still fails
   QC_TIBAE->addMethod("callOperationAsync",          (q_method_t)TIBAE_operationsAsyncCall);
   QC_TIBAE->addMethod("getAsyncOperationCallResult", (q_method_t)TIBAE_operationsGetAsyncCallResult);
   */

   return QC_TIBAE;
}

